src/CommandLine.h

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_COMMANDLINE_H
+#define SQUID_COMMANDLINE_H
+
+#if HAVE_GETOPT_H
+#include <getopt.h>
+#endif
+#include <vector>
+
+typedef struct option LongOption;
+
+/// A struct option C++ wrapper, helps with option::name copying/freeing.
+class Option : public LongOption
+{
+public:
+    Option();
+    explicit Option(const LongOption &);
+    Option(const Option&);
+    Option &operator =(const Option &);
+    ~Option();
+
+private:
+    void copy(const LongOption &);
+};
+
+/// Manages arguments passed to a program (i.e., main(argc, argv) parameters).
+class CommandLine
+{
+public:
+    /// expects main() input plus getopt_long(3) grammar rules for parsing argv
+    CommandLine(int argc, char *argv[], const char &shortRules, const LongOption *longRules);
```

Comment:
I find it very strange to pass c-strings as a `char` reference. The advantages of doing so (making sure the parameter is not nil, I presume?) do not outweigh the disadvantages of such a huge deviation from the norm. If you want to make sure that short rules are never nil when passed to getopt() and such, please check them inside the constructor and replace nil short options with `""` (an empty c-string) if needed.

Reply:
Done.